```
Error running mint automation
+ mkdir -p 12424-ffb9899/gopath/src/github.com/minio
+ git -C 12424-ffb9899/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 12424-ffb9899/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 12424-ffb9899/gopath/src/github.com/minio/minio fetch --quiet minio pull/12424/head:pr12424
+ git -C 12424-ffb9899/gopath/src/github.com/minio/minio checkout --quiet pr12424
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/12424-ffb9899/gopath
+ make -C 12424-ffb9899/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/minio/cli v1.22.0
go: downloading github.com/minio/pkg v1.0.4
go: downloading github.com/minio/minio-go/v7 v7.0.11-0.20210302210017-6ae69c73ce78
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading cloud.google.com/go v0.65.0
go: downloading google.golang.org/api v0.31.0
go: downloading github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading github.com/minio/madmin-go v1.0.10-0.20210602195449-b1bf23ec13e4
go: downloading github.com/colinmarc/hdfs/v2 v2.2.0
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/json-iterator/go v1.1.11
go: downloading github.com/fatih/color v1.12.0
go: downloading github.com/montanaflynn/stats v0.5.0
go: downloading github.com/minio/kes v0.14.0
go: downloading golang.org/x/crypto v0.0.0-20210421170649-83a5a9bb288b
go: downloading golang.org/x/sys v0.0.0-20210510120138-977fb7262007
go: downloading github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/google/uuid v1.1.2
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/cespare/xxhash/v2 v2.1.1
go: downloading github.com/gorilla/handlers v1.5.1
go: downloading github.com/VividCortex/ewma v1.1.1
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/miekg/dns v1.1.35
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/klauspost/reedsolomon v1.9.11
go: downloading github.com/klauspost/readahead v1.3.1
go: downloading github.com/klauspost/compress v1.12.2
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/minio/rpc v1.0.0
go: downloading github.com/minio/selfupdate v0.3.1
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/philhofer/fwd v1.1.1
go: downloading github.com/pierrec/lz4 v2.6.0+incompatible
go: downloading github.com/prometheus/client_golang v1.8.0
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/procfs v0.6.0
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/tinylib/msgp v1.1.6-0.20210521143832-0becd170c402
go: downloading go.etcd.io/etcd v0.0.0-20191023171146-3cf2f69b5738
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/coredns/coredns v1.4.0
go: downloading go.uber.org/zap v1.16.1-0.20210329175301-c23abee72d19
go: downloading github.com/minio/simdjson-go v0.2.1
go: downloading github.com/Shopify/sarama v1.27.2
go: downloading github.com/eclipse/paho.mqtt.golang v1.3.0
go: downloading github.com/elastic/go-elasticsearch/v7 v7.12.0
go: downloading github.com/go-sql-driver/mysql v1.5.0
go: downloading github.com/gomodule/redigo v2.0.0+incompatible
go: downloading github.com/lib/pq v1.9.0
go: downloading github.com/nats-io/nats.go v1.10.0
go: downloading github.com/nats-io/stan.go v0.8.3
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/olivere/elastic/v7 v7.0.22
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/go-ldap/ldap v3.0.2+incompatible
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/minio/parquet-go v1.0.0
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/mattn/go-ieproxy v0.0.0-20190702010315-6dee0af9227d
go: downloading github.com/mattn/go-colorable v0.1.8
go: downloading github.com/mattn/go-isatty v0.0.13
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/felixge/httpsnoop v1.0.1
go: downloading github.com/klauspost/cpuid v1.3.1
go: downloading golang.org/x/net v0.0.0-20210421230115-4e50805a0758
go: downloading github.com/golang/snappy v0.0.3
go: downloading github.com/mattn/go-runewidth v0.0.7
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/gorilla/websocket v1.4.2
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading github.com/go-ldap/ldap/v3 v3.2.4
go: downloading golang.org/x/text v0.3.6
go: downloading git.apache.org/thrift.git v0.13.0
go: downloading github.com/tidwall/sjson v1.1.6
go: downloading github.com/tidwall/gjson v1.7.5
go: downloading github.com/modern-go/reflect2 v1.0.1
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/nats-io/nkeys v0.1.4
go: downloading github.com/nats-io/jwt v1.1.0
go: downloading go.uber.org/atomic v1.7.0
go: downloading go.uber.org/multierr v1.6.0
go: downloading github.com/minio/argon2 v1.0.0
go: downloading github.com/shirou/gopsutil/v3 v3.21.4
go: downloading github.com/prometheus/common v0.14.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading github.com/lestrrat-go/jwx v1.2.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20200615164410-66371956d46c
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.1
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/tidwall/pretty v1.1.0
go: downloading github.com/tidwall/match v1.0.3
go: downloading google.golang.org/protobuf v1.26.0
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/go-resiliency v1.2.0
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading gopkg.in/jcmturner/gokrb5.v7 v7.5.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20200313005456-10cdbea86bc0
go: downloading gopkg.in/ini.v1 v1.57.0
go: downloading github.com/minio/md5-simd v1.1.0
go: downloading github.com/rs/xid v1.2.1
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1
go: downloading github.com/klauspost/cpuid/v2 v2.0.4
go: downloading github.com/tklauser/go-sysconf v0.3.5
go: downloading golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
go: downloading google.golang.org/grpc v1.29.1
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.7
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading github.com/tklauser/numcpus v0.2.2
go: downloading google.golang.org/genproto v0.0.0-20200901141002-b3bf27a9dbd1
go: downloading github.com/googleapis/gax-go/v2 v2.0.5
go: downloading go.opencensus.io v0.22.5
go: downloading github.com/jcmturner/goidentity/v6 v6.0.1
go: downloading github.com/mailru/easyjson v0.7.6
go: downloading github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
go: downloading github.com/josharian/intern v1.0.0
go: downloading go.etcd.io/etcd/api/v3 v3.5.0-beta.4
go: downloading go.etcd.io/etcd/client/v3 v3.5.0-beta.4
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.0-beta.4
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/coreos/go-systemd v0.0.0-20180511133405-39ca1b05acc7
go: downloading github.com/coreos/go-systemd/v22 v22.3.1
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: downloading gopkg.in/jcmturner/rpc.v1 v1.1.0
# github.com/minio/minio/cmd
cmd/common-main.go:97:27: syntax error: unexpected {, expecting expression
cmd/common-main.go:100:40: syntax error: unexpected ) at end of statement
cmd/common-main.go:101:5: syntax error: unexpected comma after top level declaration
make: *** [build] Error 2

```

--- a/cmd/common-main.go
+++ b/cmd/common-main.go
@@ -27,6 +27,7 @@ import (
 	"fmt"
 	"math/rand"
 	"net"
+	"net/http"
 	"net/url"
 	"os"
 	"path/filepath"
@@ -39,6 +40,7 @@ import (
 	dns2 "github.com/miekg/dns"
 	"github.com/minio/cli"
 	"github.com/minio/kes"
+	"github.com/minio/minio-go/v7/pkg/credentials"
 	"github.com/minio/minio-go/v7/pkg/set"
 	"github.com/minio/minio/internal/auth"
 	"github.com/minio/minio/internal/config"
@@ -54,6 +56,7 @@ import (
 
 // serverDebugLog will enable debug printing
 var serverDebugLog = env.Get("_MINIO_SERVER_DEBUG", config.EnableOff) == config.EnableOn
+var defaultAWSCredProvider []credentials.Provider
 
 func init() {
 	rand.Seed(time.Now().UTC().UnixNano())
@@ -74,7 +77,6 @@ func init() {
 		// safe to assume a higher timeout upto 10 minutes.
 		globalDNSCache = xhttp.NewDNSCache(10*time.Minute, 5*time.Second, logger.LogOnceIf)
 	}
-
 	initGlobalContext()
 
 	globalForwarder = handlers.NewForwarder(&handlers.Forwarder{
@@ -92,6 +94,14 @@ func init() {
 	console.SetColor("Debug", color.New())
 
 	gob.Register(StorageErr(""))
+
+	defaultAWSCredProvider = []credentials.Provider{
+		&credentials.IAM{
+			Client: &http.Client{
+				Transport: NewGatewayHTTPTransport(),
+			},
+		},
+	}
 }
 
 func verifyObjectLayerFeatures(name string, objAPI ObjectLayer) {

--- a/cmd/tier.go
+++ b/cmd/tier.go
@@ -144,12 +144,16 @@ func (config *TierConfigMgr) Edit(ctx context.Context, tierName string, creds ma
 	newCfg := config.Tiers[tierName]
 	switch tierType {
 	case madmin.S3:
-		if creds.AccessKey == "" || creds.SecretKey == "" {
+		if (creds.AccessKey == "" || creds.SecretKey == "") && !creds.AWSRole {
 			return errTierInsufficientCreds
 		}
-		newCfg.S3.AccessKey = creds.AccessKey
-		newCfg.S3.SecretKey = creds.SecretKey
-
+		switch {
+		case creds.AWSRole:
+			newCfg.S3.AWSRole = true
+		default:
+			newCfg.S3.AccessKey = creds.AccessKey
+			newCfg.S3.SecretKey = creds.SecretKey
+		}
 	case madmin.Azure:
 		if creds.AccessKey == "" || creds.SecretKey == "" {
 			return errTierInsufficientCreds

--- a/cmd/warm-backend-s3.go
+++ b/cmd/warm-backend-s3.go
@@ -99,7 +99,12 @@ func newWarmBackendS3(conf madmin.TierS3) (*warmBackendS3, error) {
 	if err != nil {
 		return nil, err
 	}
-	creds := credentials.NewStaticV4(conf.AccessKey, conf.SecretKey, "")
+	var creds *credentials.Credentials
+	if conf.AWSRole {
+		creds = credentials.NewChainCredentials(defaultAWSCredProvider)
+	} else {
+		creds = credentials.NewStaticV4(conf.AccessKey, conf.SecretKey, "")
+	}
 	getRemoteTargetInstanceTransportOnce.Do(func() {
 		getRemoteTargetInstanceTransport = newGatewayHTTPTransport(10 * time.Minute)
 	})

--- a/docs/bucket/lifecycle/README.md
+++ b/docs/bucket/lifecycle/README.md
@@ -116,6 +116,11 @@ Using above tier, set up a lifecycle rule with transition:
  mc ilm add --expiry-days 365 --transition-days 45 --storage-class "AZURETIER" myminio/srcbucket
 ```
 
+Note: In the case of S3, it is possible to create a tier from MinIO running in EC2 to S3 using AWS role attached to EC2 as credentials instead of accesskey/secretkey:
+```
+mc admin tier add s3 source S3TIER --bucket s3bucket --prefix testprefix/ --use-aws-role
+```
+
 Once transitioned, GET or HEAD on the object will stream the content from the transitioned tier. In the event that the object needs to be restored temporarily to the local cluster, the AWS [RestoreObject API](https://docs.aws.amazon.com/AmazonS3/latest/API/API_RestoreObject.html) can be utilized.
 
 ```

--- a/go.mod
+++ b/go.mod
@@ -44,7 +44,7 @@ require (
 	github.com/minio/csvparser v1.0.0
 	github.com/minio/highwayhash v1.0.2
 	github.com/minio/kes v0.14.0
-	github.com/minio/madmin-go v1.0.9
+	github.com/minio/madmin-go v1.0.10-0.20210602195449-b1bf23ec13e4
 	github.com/minio/minio-go/v7 v7.0.11-0.20210302210017-6ae69c73ce78
 	github.com/minio/parquet-go v1.0.0
 	github.com/minio/pkg v1.0.4

